File: pnix-cljs/cljs-meta/test/independent_mini_backend_test.js
const assert = require("node:assert/strict");
const cljsMeta = require("../dist/cljs-meta-module.js");
const mini = require("../independent_mini_backend.js");

// Trusting-Trust (Diverse Double-Compiling) witness: cross-check the
// self-hosted cljs.js-backed compiler against a from-scratch, independent
// tokenizer/reader + direct JS-text emitter (independent_mini_backend.js)
// that shares no code with cljs.js/cljs.compiler/cljs.analyzer. See
// STATUS.md's "Trusting-Trust defense roadmap" for the honest scope: this is
// a bounded fixture subset, not full ClojureScript, and behavior equivalence
// (not bit-identical JS text) is the bar.
const FIXTURES = [
  ["(let [x 20] (+ x 22))", 42],
  ["(if true :yes :no)", "yes"],
  ["(let [x 5 y 7] (if (< x y) (* (+ x 1) y) (- x y)))", 42],
  ["(+ 1 2)", 3],
  ["(- 50 8)", 42],
  ["(* 6 7)", 42],
  ["(if false 0 42)", 42],
  ["(let [x 41] (+ x 1))", 42],
  ['"hello"', "hello"],
  ["(do 1 2 3)", 3],
  ["(let [a 1 b 2] [a b])", [1, 2]],
  ["((fn [x] (* x x)) 6)", 36],
  ["((fn fact [n] (if (<= n 1) 1 (* n (fact (- n 1))))) 6)", 720],
  ["((fn fib [n] (if (< n 2) n (+ (fib (- n 1)) (fib (- n 2))))) 10)", 55],
  ["{:a 1 :b 2}", { a: 1, b: 2 }],
  ["(get {:a 1 :b 2} :a)", 1],
  ["(count [1 2 3])", 3],
  ["(conj [1 2] 3)", [1, 2, 3]],
  ["(nth [10 20 30] 1)", 20],
  ["(let [[a b] [20 22]] (+ a b))", 42],
  ["(let [[a b c] [20 22]] (if (nil? c) (+ a b) 0))", 42],
  ["(let [[[a b] c] [[1 2] 3]] (+ (+ a b) c))", 6],
  ["(let [[a [b c] d] [1 [2 3] 4]] (+ (+ a b) (+ c d)))", 10],
  ["#{1 2 3}", [1, 2, 3]],
  ["(count #{1 2 3})", 3],
  ["(assoc {:a 1} :b 2)", { a: 1, b: 2 }],
  ["(assoc {:a 1} :b 2 :c 3)", { a: 1, b: 2, c: 3 }],
  ["(update {:a 1} :a (fn [x] (+ x 1)))", { a: 2 }],
  ["(when true 42)", 42],
  ["(when false 42)", null],
  ["(when (> 5 3) (+ 1 2))", 3],
  ["(cond false 1 true 2 :else 3)", 2],
  ["(cond false 1 false 2 :else 99)", 99],
  ["(-> 1 (+ 2) (* 3))", 9],
  ["(loop [i 0 acc 0] (if (< i 10) (recur (+ i 1) (+ acc i)) acc))", 45],
  ["(loop [i 5 acc 1] (if (<= i 1) acc (recur (- i 1) (* acc i))))", 120],
  ["(loop [a 1 b 2 n 3] (if (= n 0) a (recur b a (- n 1))))", 2],
  ["(let [square (fn [x] (* x x))] (+ (square 3) (square 4)))", 25],
  ["(let [addA (fn [x] (+ x 5))] (- (addA 10) (addA 1)))", 9],
  ["(let [addAB (fn [a b] (+ a b))] (addAB 14 (addAB 14 14)))", 42],
  [
    "(let [base (fn [x] (+ x 3)) scaled (fn [y] (* (base y) 2))] (+ (scaled 1) (scaled 2)))",
    18,
  ],
];

(async () => {
  for (const [source, expected] of FIXTURES) {
    const hostResult = await cljsMeta.evaluate(source);
    assert.equal(hostResult.outcome_kind, "done", `host failed on: ${source}`);
    // deepEqual (not equal/strictEqual): some fixtures return vectors, which
    // cljs.js and the mini backend each produce as freshly-allocated JS
    // arrays, so reference equality would spuriously fail structurally-equal
    // results.
    assert.deepEqual(hostResult.value, expected, `host mismatch on: ${source}`);

    const miniResult = mini.compileAndEval(source);
    assert.deepEqual(miniResult, expected, `mini backend mismatch on: ${source}`);
  }

  console.log(`independent mini backend DDC: PASS (${FIXTURES.length} fixtures)`);
})().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: pnix-clr/clr-meta/test/pnix/clr_meta/independent_mini_backend_test.clj
(ns pnix.clr-meta.independent-mini-backend-test
  "Trusting-Trust (Diverse Double-Compiling) witness: cross-check real host
  ClojureCLR `eval` against `independent-mini-backend`, a from-scratch
  reader+analyzer+DynamicMethod IL emitter that shares no code with the
  Compiler Stage1-7 family. See STATUS.md's 'Trusting-Trust defense roadmap'
  for the honest scope: a bounded fixture subset, behavior equivalence (not
  bit-identical IL), not the checked-Int64 expression profile Stage1-7
  formally closes."
  (:require [clojure.test :refer [deftest is testing]]
            [pnix.clr-meta.independent-mini-backend :as mini]))

(def ^:private fixtures
  [["(fn [x] (+ x 1))" [41] 42]
   ["(fn [x y] (if (< x y) (* (+ x 1) y) (- x y)))" [5 7] 42]
   ["(fn [] (* 6 7))" [] 42]
   ["(fn [x] (- x 1))" [43] 42]
   ["(fn [x y] (+ x y))" [20 22] 42]
   ["(fn [x] (if (> x 0) x (- 0 x)))" [-42] 42]
   ["(fn [x] (if (= x 42) 1 0))" [42] 1]
   ["(fn [x] (if (>= x 41) 42 0))" [41] 42]
   ["(fn [x] (if (> x 0) (if (> x 100) 2 1) 0))" [50] 1]
   ["(fn [x] (if (> x 0) (if (> x 100) 2 1) 0))" [200] 2]
   ["(fn [x] (if (> x 0) (if (> x 100) 2 1) 0))" [-5] 0]
   ["(fn [a b c] (+ (+ a b) c))" [10 20 12] 42]
   ["(fn [a b c] (if (< a b) (+ b c) (- b c)))" [1 20 22] 42]
   ["(fn [a b c d] (+ (+ a b) (+ c d)))" [10 10 10 12] 42]
   ["(fn [a b c d] (if (> a b) (+ c d) (- d c)))" [5 1 20 22] 42]
   ["(fn [x] (let [y (+ x 1)] (* y 2)))" [10] 22]
   ["(fn [x y] (let [a (+ x y) b (* a 2)] (- b 1)))" [3 4] 13]
   ["(fn [x] (let [y (if (> x 0) x (- 0 x))] (* y y)))" [-5] 25]
   ["(fn [x] (let [x (+ x 1)] (* x 2)))" [10] 22]
   ["(fn [x] (let [a (+ x 1)] (let [b (* a 2)] (+ a b))))" [5] 18]
   ["(fn [n] (loop [i 0 acc 0] (if (> i n) acc (recur (+ i 1) (+ acc i)))))" [10] 55]
   ["(fn [n] (loop [i n acc 1] (if (<= i 1) acc (recur (- i 1) (* acc i)))))" [5] 120]
   ["(fn [n acc] (if (= n 0) acc (recur (- n 1) (+ acc n))))" [5 0] 15]
   ["(fn [n a b] (loop [i n a a b b] (if (= i 0) a (recur (- i 1) b a))))" [3 1 2] 2]
   ["(fn [] ((fn [x] (+ x 1)) 41))" [] 42]
   ["(fn [] (((fn [x] (fn [y] (+ x y))) 20) 22))" [] 42]
   ["(fn [] ((((fn [a] (fn [b] (fn [c] (+ a (+ b c))))) 1) 2) 3))" [] 6]
   ["(fn [n] (if (> n 0) ((fn [x] (* x 2)) n) 0))" [5] 10]
   ["(fn [n] (let [square (fn [x] (* x x))] (square n)))" [7] 49]
   ["(fn [n] (let [k 10 add-k (fn [x] (+ x k))] (add-k n)))" [5] 15]])

(deftest independent-mini-backend-agrees-with-host-eval
  (testing "real host ClojureCLR eval and the independent mini backend agree"
    (doseq [[source args expected] fixtures]
      (let [host-fn (eval (read-string source))
            host-result (long (apply host-fn args))
            mini-result (long (mini/compile-and-invoke source args))]
        (is (= expected host-result) (str "host mismatch: " source))
        (is (= expected mini-result) (str "mini backend mismatch: " source))))))

(def ^:private overflow-fixtures
  "checked-i64-expression profile's :overflow :system-overflow-exception --
  both the real host and the independent mini backend use checked (not
  wrapping) Int64 arithmetic, so both must reject these the same way."
  [["(fn [x] (+ x 1))" [Int64/MaxValue]]
   ["(fn [x] (- x 1))" [Int64/MinValue]]
   ["(fn [x] (* x 2))" [Int64/MaxValue]]
   ["(fn [x y] (+ x y))" [Int64/MaxValue Int64/MaxValue]]])

(deftest independent-mini-backend-agrees-on-checked-overflow
  (testing "real host ClojureCLR eval and the independent mini backend both reject Int64 overflow"
    (doseq [[source args] overflow-fixtures]
      (let [host-fn (eval (read-string source))
            host-threw? (try (apply host-fn args) false (catch Exception _ true))
            mini-threw? (try (mini/compile-and-invoke source args) false (catch Exception _ true))]
        (is host-threw? (str "host did not reject overflow: " source))
        (is mini-threw? (str "mini backend did not reject overflow: " source))))))
